pkg/storage/fingerprint_writer.go

```diff
@@ -0,0 +1,141 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package storage
+
+import (
+	"context"
+	"hash"
+	"io"
+
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/errors"
+)
+
+// fingerprintWriter hashes every key/timestamp and value for point keys, and
+// combines their hashes via a XOR into a running aggregate.
+//
+// Range keys are not fingerprinted but instead written to a pebble SST that is
+// returned to the caller. This is because range keys do not have a stable,
+// discrete identity and so it is up to the caller to define a deterministic
+// fingerprinting scheme across all returned range keys.
+//
+// The caller must Finish() and Close() the fingerprintWriter to finalize the
+// writes to the underlying pebble SST.
+type fingerprintWriter struct {
+	hasher hash.Hash64
+
+	sstWriter *SSTWriter
+	xorAgg    *uintXorAggregate
+}
+
+// makeFingerprintWriter creates a new fingerprintWriter.
+func makeFingerprintWriter(
+	ctx context.Context, hasher hash.Hash64, cs *cluster.Settings, f io.Writer,
+) fingerprintWriter {
+	// TODO(adityamaru,dt): Once
+	// https://github.com/cockroachdb/cockroach/issues/90450 has been addressed we
+	// should write to a kvBuf instead of a Backup SST writer.
+	sstWriter := MakeBackupSSTWriter(ctx, cs, f)
+	return fingerprintWriter{
+		sstWriter: &sstWriter,
+		hasher:    hasher,
+		xorAgg:    &uintXorAggregate{},
+	}
+}
+
+type uintXorAggregate struct {
+	sum uint64
+}
+
+// add inserts one value into the running xor.
+func (a *uintXorAggregate) add(x uint64) {
+	a.sum = a.sum ^ x
+}
+
+// result returns the xor.
+func (a *uintXorAggregate) result() uint64 {
+	return a.sum
+}
+
+// Finish finalizes the underlying SSTWriter, and returns the aggregated
+// fingerprint for point keys.
+func (f *fingerprintWriter) Finish() (uint64, error) {
+	if err := f.sstWriter.Finish(); err != nil {
```

Comment:
nit: Should we do the same as `MVCCExportToSST` here and drop the SST `Finish()` if there is no data? Probably doesn't matter much though.

Reply:
done!